File: app/src/main/java/com/example/myanimedata/ui/TopRatedCharacterFragment.java
package com.example.myanimedata.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myanimedata.R;
import com.example.myanimedata.adapter.CharacterAdapter;
import com.example.myanimedata.api.ApiClient;
import com.example.myanimedata.api.ApiService;
import com.example.myanimedata.api.CharacterResponse;
import com.example.myanimedata.api.CharacterResult;
import com.example.myanimedata.databinding.FragmentTopRatedCharacterBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class TopRatedCharacterFragment extends Fragment {
    private ApiService apiService;
    private FragmentTopRatedCharacterBinding binding;
    private CharacterAdapter topCharacterAdapter;

    private final List<CharacterResult> topCharacterResults = new ArrayList<>();

    private int page = 1;

    public TopRatedCharacterFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentTopRatedCharacterBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        Retrofit retrofit = ApiClient.getClient();
        apiService = retrofit.create(ApiService.class);

        GridLayoutManager mLayoutManager = new GridLayoutManager(getContext(), 3);

        RecyclerView rvTopCharacterList = root.findViewById(R.id.rvTopCharacterList);
        topCharacterAdapter = new CharacterAdapter(topCharacterResults, getContext());

        rvTopCharacterList.setLayoutManager(mLayoutManager);
        rvTopCharacterList.setAdapter(topCharacterAdapter);
        getTopCharacterData(page);

        rvTopCharacterList.addOnScrollListener(new RecyclerView.OnScrollListener(){
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                super.onScrolled(recyclerView, dx, dy);

                if(!recyclerView.canScrollVertically(1)){
                    page++;
                    getTopCharacterData(page);
                }
            }
        });

        return root;
    }

    private void getTopCharacterData(int page) {
        int limit = 15;

        Call<CharacterResponse> call = apiService.getTopCharacter(page, limit);
        call.enqueue(new Callback<CharacterResponse>() {
            @Override
            public void onResponse(@NonNull Call<CharacterResponse> call, @NonNull Response<CharacterResponse> response) {
                if(response.body() != null){
                    int oldCount = topCharacterResults.size();
                    binding.loadingTopCharacter.setVisibility(View.GONE);
                    binding.rvTopCharacterList.setVisibility(View.VISIBLE);

                    topCharacterResults.addAll(response.body().getCharacterResults());
                    topCharacterAdapter.notifyItemRangeInserted(oldCount, topCharacterResults.size());
                } else if (topCharacterResults.isEmpty()) {
                    binding.loadingTopCharacter.setVisibility(View.GONE);
                    binding.textNoTopResult.setVisibility(View.VISIBLE);
                } else {
                    binding.loadingTopCharacter.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<CharacterResponse> call, @NonNull Throwable t) {
                binding.loadingTopCharacter.setVisibility(View.GONE);
                Toast.makeText(getContext(), t.getMessage() + " cause : "
                        + t.getCause(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/myanimedata/ui/TopRatedCharacterFragment.kt
package com.example.myanimedata.ui

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myanimedata.R
import com.example.myanimedata.adapter.CharacterAdapter
import com.example.myanimedata.api.ApiClient.client
import com.example.myanimedata.api.ApiService
import com.example.myanimedata.api.CharacterResponse
import com.example.myanimedata.api.CharacterResult
import com.example.myanimedata.databinding.FragmentTopRatedCharacterBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class TopRatedCharacterFragment : Fragment() {
    private var apiService: ApiService? = null
    private var binding: FragmentTopRatedCharacterBinding? = null

    private var topCharacterAdapter: CharacterAdapter? = null
    private val topCharacterResults: MutableList<CharacterResult> = ArrayList()

    private var page = 1

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = FragmentTopRatedCharacterBinding.inflate(inflater, container, false)

        val root: View = binding!!.root
        val retrofit = client
        apiService = retrofit!!.create(ApiService::class.java)

        val mPortraitLayoutManager = GridLayoutManager(context, 3)
        val mLandscapeLayoutManager = GridLayoutManager(context, 5)
        val rvTopCharacterList = root.findViewById<RecyclerView>(R.id.rvTopCharacterList)
        topCharacterAdapter = context?.let { CharacterAdapter(topCharacterResults, it) }

        if(resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvTopCharacterList.layoutManager = mLandscapeLayoutManager
        } else {
            rvTopCharacterList.layoutManager = mPortraitLayoutManager
        }

        rvTopCharacterList.adapter = topCharacterAdapter
        rvTopCharacterList.setHasFixedSize(true)
        getTopCharacterData(page)

        rvTopCharacterList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!recyclerView.canScrollVertically(1)) {
                    page++
                    getTopCharacterData(page)
                }
            }
        })
        return root
    }

    private fun getTopCharacterData(page: Int) {
        val limit = 15
        val call = apiService!!.getTopCharacter(page, limit)

        call!!.enqueue(object : Callback<CharacterResponse?> {
            override fun onResponse(
                call: Call<CharacterResponse?>,
                response: Response<CharacterResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = topCharacterResults.size
                    binding!!.loadingTopCharacter.visibility = View.GONE
                    binding!!.rvTopCharacterList.visibility = View.VISIBLE
                    response.body()!!.characterResults?.let { topCharacterResults.addAll(it) }
                    topCharacterAdapter!!.notifyItemRangeInserted(
                        oldCount,
                        topCharacterResults.size
                    )
                } else if (topCharacterResults.isEmpty()) {
                    binding!!.loadingTopCharacter.visibility = View.GONE
                    binding!!.textNoTopResult.visibility = View.VISIBLE
                } else {
                    binding!!.loadingTopCharacter.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<CharacterResponse?>, t: Throwable) {
                binding!!.loadingTopCharacter.visibility = View.GONE
                Toast.makeText(
                    context, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
}